### ISSUE TITLE
Document: is libmtp thread safe?

### ISSUE THREAD

**Issue by noell** [libmtp/libmtp#141]
@msmeissn https://github.com/libmtp main page does not say anything about libmtp thread-safety.  My _guess_ is libmtp is not thread safe.  Is that correct?

**Comment by msmeissn** [comment]
I checked and currently they are some global variables in use which would make it not threadsafe.

e.g. 
libusb1-glue.c:
static libusb_context *libmtp_libusb_context;
(global libusb context..)

libmtp.c:
static filemap_t *g_filemap = NULL;
static propertymap_t *g_propertymap = NULL;

(initialized once by LIBMTP_Init())

It should not be hard to remove these last thread unsafe global variables if there is need, but currently I think we cannot claim thread safety.

**Comment by noell** [comment]
Agree: libmtp is not thread-safe due to global variables.  Systems I work on (Chrome, ChromeOS, Android) only allow single-threaded access to libmtp for that reason.

> It should not be hard to remove these last thread unsafe global variables if there is need, but currently I think we cannot claim thread safety.

No real need from my point of view.  If two or more threads want to access to libmtp, they should acquire a mutex lock before calling libmtp API. That's simple to do in practice, and guarantees thread-safe access to libmtp.

**Comment by noell** [comment]
Resolve: webcrawler's will index our discussion and help the curious.  A one-line statement about libmtp thread safety on its main page would be a nice to have.  That can be done anytime.
